### PR TITLE
Add getPagination* concrete functions on Pagination trait

### PR DESCRIPTION
**Pull request by pirogoeth**

### CODE REVIEW

File: src/Mailgun/Api/Pagination.php [hunk 1]
```diff
@@ -22,6 +23,51 @@
...
+     * @return string
+     */
+    abstract protected function getPaginationBase();
+
+    /**
+     * @param PaginationResponse $response
+     *
+     * @return mixed|null
+     */
+    public function getPaginationNext(PaginationResponse $response)
```
Comment (Nyholm) [inline src/Mailgun/Api/Pagination.php]: PaginationResponse is a Trait. You cannot use Traits like this. 
https://3v4l.org/K4mdT

File: src/Mailgun/Api/Pagination.php [hunk 2]
```diff
@@ -22,6 +23,51 @@
...
+    {
+        return $this->getPaginationUrl($response->getNextUrl(), $this->getPaginationBase());
```
Comment (Nyholm) [inline src/Mailgun/Api/Pagination.php]: Why not use `get_class($response);`?

**Comment by pirogoeth** [comment]
@Nyholm addressed your comments, sorry for the delay :)

File: src/Mailgun/Api/Pagination.php [hunk 3]
```diff
@@ -10,6 +10,7 @@
 namespace Mailgun\Api;
 
 use Mailgun\Assert;
+use Mailgun\Resource\Api\mixed;
```
Comment (DavidGarciaCat) [inline src/Mailgun/Api/Pagination.php]: 👎 this is not a valid resource that we can import. If we can use more than one based on "criteria" then I suggest to add `use Mailgun\Resource\Api;` and other annotations and/or type hint can use `Api\...`

File: src/Mailgun/Api/Pagination.php [hunk 4]
```diff
@@ -22,12 +23,52 @@
     abstract protected function safeDeserialize(ResponseInterface $response, $className);
 
     /**
+     * @param mixed $response
+     *
+     * @return mixed|null
```
Comment (DavidGarciaCat) [inline src/Mailgun/Api/Pagination.php]: `mixed` type hint (s)

File: src/Mailgun/Api/Pagination.php (hunk 3 above)
Comment (DavidGarciaCat) [inline src/Mailgun/Api/Pagination.php]: Plan B is to set a generic Interface that should be implemented by all possible objects that we are going to expect.

Comment (pirogoeth) [inline src/Mailgun/Api/Pagination.php]: Whoops, didn't mean to do that! Did a find-and-replace on PaginationResponse to get rid of the type hints and set the annotations to mixed, but I guess I missed one :) nice catch!

File: src/Mailgun/Api/Pagination.php (hunk 4 above)
Comment (pirogoeth) [inline src/Mailgun/Api/Pagination.php]: Uhh, what do you mean?

File: src/Mailgun/Api/Pagination.php [hunk 5]
```diff
@@ -22,12 +22,52 @@
...
+     */
+    public function getPaginationNext($response)
```
Comment (Nyholm) [inline src/Mailgun/Api/Pagination.php]: We should still try to type hint this. 
What do you think of adding an interface that cover the `get*Url()` functions from the trait?

File: src/Mailgun/Api/Pagination.php [hunk 6]
```diff
@@ -22,12 +22,52 @@
...
+    {
+        return $this->getPaginationUrl($response->getNextUrl(), get_class($response));
+    }
+
+    /**
+     * @param mixed $response
+     *
+     * @return mixed|null
+     */
+    public function getPaginationPrevious($response)
```
Comment (Nyholm) [inline src/Mailgun/Api/Pagination.php]: Just an idea: What do you think of renaming the functions to `paginationPrevious`?

Comment (pirogoeth) [inline src/Mailgun/Api/Pagination.php]: I certainly see no problem with that.

File: src/Mailgun/Api/Pagination.php (hunk 5 above)
Comment (pirogoeth) [inline src/Mailgun/Api/Pagination.php]: Yeah, that could work quite well. I'll take care of that on Monday.

File: src/Mailgun/Api/Pagination.php (hunk 4 above)
Comment (DavidGarciaCat) [inline src/Mailgun/Api/Pagination.php]: Example:
When we call the API endpoint then the SDK should return a Response, right?

What I mean - ideally - we should type hint the expected content, and `mixed` has no sense to me when I am using a SDK that is managing an API resource. From my POV a SDK should return always known resources, no matter if it is an object, an array, a string...

File: src/Mailgun/Api/Pagination.php (hunk 6 above)
Comment (DavidGarciaCat) [inline src/Mailgun/Api/Pagination.php]: I would say `previous` or `previousPage` due we are already using a Pagination object does not make sense to "persist" using `pagination...`

Comment (pirogoeth) [inline src/Mailgun/Api/Pagination.php]: Or even this ^^.

I like just being able to do `$resp->previous()` and so on, as it's very clean and not redundant.

Comment (DavidGarciaCat) [inline src/Mailgun/Api/Pagination.php]: Agree

Comment (Nyholm) [inline src/Mailgun/Api/Pagination.php]: Are we sure that no API will contain ever a call called "next", "previus", "first" and "last"?

Comment (DavidGarciaCat) [inline src/Mailgun/Api/Pagination.php]: Well, that is why I suggested `previousPage()` for example, but in the other hand, we are working with a Pagination object so make sense to use a method name like `previous()`. Feel free to choose, as far it works as expected.

Comment (pirogoeth) [inline src/Mailgun/Api/Pagination.php]: I am renaming to `(previous|next|first|last)Page`.

**Pull request merged**